wire-schema/src/main/java/com/squareup/wire/schema/SemVer.kt

```diff
@@ -0,0 +1,128 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.schema
+
+import java.util.Locale
+
+/**
+ * A version string as specified by [semver.org][semver_org]. This is used to order versions for
+ * sorting since and until.
+ *
+ * If there is a `+` in the version string it delimits the build metadata. The plus character and
+ * the section that follows are unused for version comparisons.
+ *
+ * The rest of the string is split into two sections:
+ *
+ *  * Release (preceding the first dash). Typically like "1.0.0".
+ *
+ *  * Pre-release (optional, following the first dash). Typically like "alpha", "snapshot", or
+ *    "rc1".
+ *
+ * Each section has one or more segments separated by dots and dashes. Within a segment comparisons
+ * are either numeric (if both segments are strictly digits), lexicographic (if both contain
+ * non-digit characters) or numbers-first (if only one segment is strictly digits).
+ *
+ * This class requires lowercase versions to defend against potentially surprising behavior in the
+ * semver.org spec ("1.0-alpha" comes after "1.0-BETA" due to case-sensitive sorting).
+ *
+ * [semver_org]: https://semver.org/
+ */
+internal data class SemVer(val version: String) : Comparable<SemVer> {
+
+  init {
+    require(version.toLowerCase(Locale.US) == version) { "version must be lowercase: $version" }
+  }
+
+  override fun compareTo(other: SemVer): Int {
+    return compareSegments(version, other.version) { a, b ->
+      val aDigits = a.toBigIntegerOrNull()
+      val bDigits = b.toBigIntegerOrNull()
+      return@compareSegments when {
+        aDigits != null && bDigits != null -> aDigits.compareTo(bDigits)
+        aDigits != null -> -1
+        bDigits != null -> 1
+        else -> a.compareTo(b)
+      }
+    }
+  }
+
+  override fun toString() = version
+
+  private fun compareSegments(a: String, b: String, comparator: SegmentComparator): Int {
+    var aPos = 0
+    var bPos = 0
+
+    for (terminators in listOf(RELEASE_TERMINATORS, PRERELEASE_TERMINATORS)) {
+      val aSize = a.find(terminators, startIndex = aPos)
+      val bSize = b.find(terminators, startIndex = bPos)
+
+      // A version that lacks a pre-release section comes after a version that has one.
+      if (terminators === PRERELEASE_TERMINATORS) {
+        val aNoPrerelease = aPos == aSize
+        val bNoPrerelease = bPos == bSize
+        if (aNoPrerelease && !bNoPrerelease) return 1
+        if (bNoPrerelease && !aNoPrerelease) return -1
+      }
+
+      while (aPos < aSize && bPos < bSize) {
+        val aLimit = a.find(SEPARATORS, startIndex = aPos, endIndex = aSize)
+        val bLimit = b.find(SEPARATORS, startIndex = bPos, endIndex = bSize)
+
+        val result = comparator(
```

Comment:
What's this do? I would expect a call to `compareTo`.

Reply:
I’m inadvertently using a lambda instead of a proper comparator. lemme roll that back.